vispy/visuals/volume.py

```diff
@@ -506,13 +522,75 @@ def set_data(self, vol, clim=None, copy=True):
         # Get some stats
         self._kb_for_texture = np.prod(self._vol_shape) / 1024
     
+    def rescale_data(self):
+        """Force rescaling of data to the current contrast limits and texture upload.
+
+        Because Textures are currently 8-bits, and contrast adjustment is done during
+        rendering by scaling the values retrieved from the texture on the GPU (provided that
+        the new contrast limits settings are within the range of the clims used when the
+        last texture was uploaded), posterization may become visible if the contrast limits
+        become *too* small of a fraction of the clims used to normalize the texture.
+        This function is a convenience to "force" rescaling of the Texture data to the
+        current contrast limits range.
+        """
+        self.set_data(self._last_data, clim=self._clim)
+        self.update()
+
     @property
     def clim(self):
-        """ The contrast limits that were applied to the volume data.
-        Settable via set_data().
+        """Contrast Limits. Volume display is mapped from black to white with these values.
+        Settable via set_data() as well as @clim.setter.
         """
         return self._clim
-    
+
+    @clim.setter
+    def clim(self, value):
+        """Set contrast limits used when rendering the image.
+
+        ``value`` should be a 2-tuple of floats (min_clim, max_clim), where each value is
+        within the range set by self.clim. If the new value is outside of the (min, max)
+        range of the clims previously used to normalize the texture data, then data will
+        be renormalized using set_data.
+        """
+        clim = np.array(value, float)
+        if not (clim.ndim == 1 and clim.size == 2):
+            raise ValueError('clim must be a 2-element array-like')
+        self._clim = tuple(clim)
+        if (clim[0] < self._texture_limits[0]) or (clim[1] > self._texture_limits[1]):
```

Comment:
I'm realizing something and I know this is probably too late to change this given you've made the ImageVisual changes already, but one problem with this approach is that if I make the color limits smaller I probably want a better color/bit depth of my image. This current approach if initialized with large-ish values and then changed to smaller limits would not update the texture and could result in only a few colors being shown on the screen.

I'm wondering if there should be an overall flag in `__init__` that controls this behavior. Like maybe it defaults to the optimized method that you've done here, but then can be turned off for "optimized" colors.

Reply:
Yeah that's what I was describing in https://github.com/vispy/vispy/pull/1842#issuecomment-600901420.
One potential solution is to have a threshold that also causes an automatic recalculation, perhaps if the clim range gets to lower than 25% of the original range? (that would correspond to ~64 colors which qualitatively feels to be about where obvious posterization starts to appear).

something like this in `clim.setter`
```python

if (clim[0] < self._texture_limits[0]) or (clim[1] > self._texture_limits[1]):
    self.rescale_data()
    return

range_ratio = np.subtract(*clim) / np.subtract(*self._texture_limits)
if np.abs(range_ratio) < self._min_clim_range_ratio:
    self.rescale_data()
else:
   ...
```

Reply:
I think this would be overkill. At this point I'm worried we're (you're) doing all this work just for a small optimization, but if the napari folks think this will improve performance then I suppose it is worth it.

Reply:
I can definitely vouch for the performance improvement when changing the contrast limits on volumes.  for volumes larger than about 512x512x256 (pretty common in microscopy) I see ~5fold improvement in FPS on my system (some numbers [here](https://github.com/napari/napari/pull/1056)).  I'm actually very excited to eventually take advantage of these two PRs.

I was already implementing something like this in the napari code to prevent posterization, so unless you're worried about it being overkill from a perspective of code cleanliness or performance in vispy, I definitely don't mind doing it.  (but if you think it's confusing or weird, that's another story! 😄)

On the other hand, if you'd rather keep it opt-in-or-nothing, can you clarify what you meant with "it defaults to the optimized method that you've done here, but then can be turned off for "optimized" colors."  Do you mean someone would be able to opt out of _any_ rescaling (whether it by upscaling or downscaling), or that they would opt in/out of rescaling if the clim range got dangerously low?  If the latter, `_min_clim_range_ratio` could be that opt-in/out value: if it was 0, no low-end rescaling would ever occur.

Anyway, I'm happy to follow your lead here!

Reply:
Sounds good. As for "optimized" colors, I meant that the person could have it always rescale or do your current "smart" rescaling. You are write that the clim range ratio could be this flag (`1` means always rescale?).

Reply:
sure!  yeah with the `range_ratio` logic above, `0` would mean never rescale, `1` would mean always rescale, and anything in between would only rescale if the current clim `range_ratio` was smaller  than `_min_clim_range_ratio`.  (and of course, it will always rescale if the new clims are outside of the previous texture limits)

Reply:
and fwiw, a "future goal" is definitely to have another look at full float support for textures, in which case we can just kinda avoid all of this right?

Reply:
For napari, yes. For people targeting older GPUs, no, they'll still want to use 8-bit textures.

Reply:
yeah def.  and I'm sure we have plenty of CPU/old-GPU using users as well, so I recognize it doesn't remove the need for this approach too.